Propagate CUDAOutOfMemoryError to Python.

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #83146

The intention is to make it easier to catch this situation for debugging,
logging, or application-specific recovery.

zdevito committed:
Propagate CUDAOutOfMemoryError to Python.

The intention is to make it easier to catch this situation for debugging,
logging, or application-specific recovery.

[ghstack-poisoned]

albanD commented on torch/csrc/Exceptions.cpp:
```diff
@@ -51,6 +52,14 @@ could not be completed because the input matrix is singular.",
 
+  ASSERT_TRUE(
+      THPException_CUDAOutOfMemoryError = PyErr_NewException(
+          "torch.cuda.CUDAOutOfMemoryError", nullptr, nullptr));
```
Should this inherit from RuntimeError or is the default Exception ok?

zdevito replied:
For bc it probably should.

albanD reviewed (approved):
Nice cleanup of the Exceptions code. Thanks!

zdevito committed:
Update on "Propagate CUDAOutOfMemoryError to Python."


The intention is to make it easier to catch this situation for debugging,
logging, or application-specific recovery.

[ghstack-poisoned]

zdevito commented:
Switching the name of the C++ exception to `c10::OutOfMemoryError` (and `torch.cuda.OutOfMemoryError`) because having it be accessible outside of CUDA confuses the hipify process. This complexity may be why it wasn't exposed before.

albanD commented on c10/util/Exception.h:
```diff
@@ -235,6 +235,10 @@ class C10_API LinAlgError : public Error {
   using Error::Error;
 };
 
+class C10_API OutOfMemoryError : public Error {
```
If we want to keep the cuda there, we could put this in a `torch::cuda::` namespace?
But on the other hand, maybe we also want our CPU allocator to raise this?

zdevito replied:
I think it is ok to generalize this. Previous to now it really was only used in some C++ tests to check that OOMs were being raised in certain situations. I can see a future where we extend the exception with a device attribute to describe which thing ran out of memory if we really needed the distinction.

albanD replied:
Sounds good!

albanD reviewed (commented):
Rename sounds ok even though it is a generalization that I'm not so sure about.

zdevito committed:
Update on "Propagate CUDAOutOfMemoryError to Python."


The intention is to make it easier to catch this situation for debugging,
logging, or application-specific recovery.

[ghstack-poisoned]

zdevito committed:
Update on "Propagate CUDAOutOfMemoryError to Python."


The intention is to make it easier to catch this situation for debugging,
logging, or application-specific recovery.

[ghstack-poisoned]

albanD reviewed (approved):
LGTM

zdevito commented:
@pytorchbot merge

zdevito committed:
Update on "Propagate CUDAOutOfMemoryError to Python."


The intention is to make it easier to catch this situation for debugging,
logging, or application-specific recovery.

[ghstack-poisoned]

zdevito commented:
@pytorchbot merge
